search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+                         char                    **results,
+                         gpointer user_data)
+{
+    g_dbus_method_invocation_return_value (invocation,
+                                          get_result_metas (results, user_data));
+}
+
+static void
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
+                      GDBusMethodInvocation              *invocation,
+                      gchar                             **terms,
+                      guint32                             timestamp,
+                      gpointer                            user_data)
+{
+    GDBusConnection *session_bus;
+    GVariant *reply;
+    GError *error = NULL;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder param_builder;
+    gchar *joined_terms = g_strjoinv (" ", terms);
+    gint i;
+
+    session_bus = g_dbus_object_manager_server_get_connection (app->object_manager);
+    if (session_bus == NULL) {
+      g_dbus_method_invocation_return_error_literal (invocation, G_IO_ERROR, G_IO_ERROR_FAILED, "failed to get connection to the session bus");
+      return;
+    }
+
+    g_variant_builder_init (&param_builder, G_VARIANT_TYPE ("av"));
+    for (i = 0; terms[i] != NULL; i++)
+        g_variant_builder_add (&param_builder, "v", g_variant_new_string (joined_terms));
+
+    reply = g_dbus_connection_call_sync (session_bus,
+                                         "org.gnome.Yelp",
+                                         "/org/gnome/Yelp",
+                                         "org.freedesktop.Application",
+                                         "ActivateAction",
+                                         g_variant_new ("(sava{sv})",
+                                                        "show-search",
+                                                        &param_builder,
+                                                        NULL),
+                                         NULL,
+                                         G_DBUS_CALL_FLAGS_NONE,
+                                         -1,
+                                         NULL,
+                                         &error);
+
+    g_object_unref (session_bus);
+    if (reply)
+        g_variant_unref (reply);
+    if (error) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return;
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+}
+
+static void
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation              *invocation,
+                        gchar                              *result,
+                        gchar                             **terms,
+                        guint32                             timestamp,
+                        gpointer                            user_data)
+{
+    GDBusConnection *session_bus;
```

Comment:
Here you seem to do similar steps as in the `handle_launch_search` function. The only difference is the action name and the contents of the GVariantBuilder. How about refactoring this a bit to avoid repetition?